### PR TITLE
Aggregate and/or Split Quotations into multiple Sales Orders 

### PR DESCRIPTION
**Pull request by fproldan**
#55

### CODE REVIEW

**Comment by agritheory** [comment]
@fproldan This looks good, but I think we need a few changes:

The 'split qty ' column should be editable
image

I think you send me a note about this in Slack, but the multi-company workflow for this does not appear to be working correctly. 

<details>

image


### App Versions
```
{
	"erpnext": "14.34.0",
	"frappe": "14.44.0",
	"hrms": "14.10.0",
	"inventory_tools": "14.3.0"
}
```
### Route
```
query-report/Quotation Demand
```
### Traceback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 94, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 47, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 85, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1620, in call
    return fn(*args, **newargs)
  File "apps/inventory_tools/inventory_tools/inventory_tools/report/quotation_demand/quotation_demand.py", line 179, in create
    so.save()
  File "apps/frappe/frappe/model/document.py", line 305, in save
    return self._save(*args, **kwargs)
  File "apps/frappe/frappe/model/document.py", line 327, in _save
    return self.insert()
  File "apps/frappe/frappe/model/document.py", line 259, in insert
    self.run_before_save_methods()
  File "apps/frappe/frappe/model/document.py", line 1051, in run_before_save_methods
    self.run_method("validate")
  File "apps/frappe/frappe/model/document.py", line 915, in run_method
    out = Document.hook(fn)(self, *args, **kwargs)
  File "apps/frappe/frappe/model/document.py", line 1277, in composer
    return composed(self, method, *args, **kwargs)
  File "apps/frappe/frappe/model/document.py", line 1259, in runner
    add_to_return_value(self, fn(self, *args, **kwargs))
  File "apps/frappe/frappe/model/document.py", line 912, in fn
    return method_object(*args, **kwargs)
  File "apps/erpnext/erpnext/selling/doctype/sales_order/sales_order.py", line 48, in validate
    super(SalesOrder, self).validate()
  File "apps/erpnext/erpnext/controllers/selling_controller.py", line 32, in validate
    super(SellingController, self).validate()
  File "apps/erpnext/erpnext/controllers/stock_controller.py", line 45, in validate
    super(StockController, self).validate()
  File "apps/erpnext/erpnext/controllers/accounts_controller.py", line 166, in validate
    self.set_total_in_words()
  File "apps/erpnext/erpnext/controllers/selling_controller.py", line 131, in set_total_in_words
    base_amount = abs(
TypeError: bad operand type for abs(): 'NoneType'

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"company": "Ambrosia Pie Company",
		"filters": "{\"end_date\":\"2024-04-15\"}",
		"rows": "[{\"quotation_item\":\"ab9cb455f3\",\"quotation\":\"SAL-QTN-2024-00004\",\"company\":\"Chelsea Fruit Co\",\"customer\":\"Almacs Food Group\",\"transaction_date\":\"2024-04-15\",\"item_code\":\"Ambrosia Pie\",\"item_name\":\"Ambrosia Pie\",\"qty\":5,\"uom\":\"Nos\",\"warehouse\":\"Finished Goods - CFC\",\"split_qty\":5,\"indent\":1},{\"quotation_item\":\"93cf598ceb\",\"quotation\":\"SAL-QTN-2024-00003\",\"company\":\"Ambrosia Pie Company\",\"customer\":\"Downtown Deli\",\"transaction_date\":\"2024-04-15\",\"item_code\":\"Ambrosia Pie\",\"item_name\":\"Ambrosia Pie\",\"qty\":2,\"uom\":\"Nos\",\"warehouse\":\"Refrigerated Display - APC\",\"split_qty\":2,\"indent\":1}]"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/inventory_tools.inventory_tools.report.quotation_demand.quotation_demand.create"
}
```
### Response Data
```
{
	"exc_type": "DoesNotExistError",
	"exception": "TypeError: bad operand type for abs(): 'NoneType'"
}
```

</details>

Also, we need documentation for this feature.

**Comment by fproldan** [comment]
@agritheory Sorry I forget to set the PR in Draft.
I'm working on the tests and fixing some issues.
The report has the 'Split Qty' column editable, I'll check if there is an issue with the versions.

**Review by agritheory — CHANGES_REQUESTED** [review]
OK, this is so close. I think we want to adapt the the draft and total selected columns from the material demand report as well. 
image

The docs look great and read great but adding those columns will change a couple of screen shots.

In my testing the SO created didn't have any pricing, which I think means we're not forwarding the quoted price to the SO. We want to preserve that I think. We should be showing the price and amount column in this report too. 

After setting up the aggregation company in Inventory Tools settings, it still created two Sales Orders. for the company-specific amounts. 

image

**Comment by agritheory** [comment]
@fproldan I just found a few more things after running the tests locally:
- Because of the "forward price from quotation" issued noted above, there should be an assert for at least one expected price
- I see the assert for warehouse in the `test_report_with_aggregation_and_aggregation_warehouse` test, lets add asserts in the other tests to make sure it isn't accidentally aggregating when it should not be.

File: inventory_tools/inventory_tools/report/quotation_demand/quotation_demand.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+		rows = list(_rows)
+		output.append({"customer": customer, "indent": 0})
+		for r in rows:
+			r.split_qty = r["qty"]
+			r.price = fmt_money(r.get("rate"), 2, r.get("currency")).replace(" ", "")
+			r.amount = fmt_money(r.get("amount"), 2, r.get("currency")).replace(" ", "")
+			r.draft_so = frappe.db.get_value(
+				"Sales Order Item",
+				{"quotation_item": r.quotation_item, "docstatus": 0},
+				"sum(qty) as qty",
```
Comment (fproldan) [inline inventory_tools/inventory_tools/report/quotation_demand/quotation_demand.py:62]: @agritheory the `sum(qty) as qty` it's something maybe we want to add to Material Demand, it handles the cases of multiples SO created based the same Quotation.

Comment (agritheory) [reply inventory_tools/inventory_tools/report/quotation_demand/quotation_demand.py:62]: @fproldan Agreed, I'll add a ticket.

**Comment by fproldan** [comment]
@agritheory this is ready for review, I'll update the docs if the issues were addressed.

**Review by agritheory — APPROVED** [review]

**Pull request merged**